sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+= sycl_ext_oneapi_async_memory_alloc
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2024 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties]
+* link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions]
+* link:../supported/sycl_ext_oneapi_default_context.asciidoc[
+  sycl_ext_oneapi_default_context]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Backend support status
+
+This extension is not yet implemented in {dpcpp}.
+
+== Overview
+
+This extension introduces an interface for allocating and freeing USM memory
+asynchronously, as well as introducing a USM memory pool from which this memory
+is allocated and freed from.
+
+The asynchornous memory allocation and free are exposed via new free functions
+taking a queue, which enqueue an asynchronous malloc or free command and return
+immediately. The functions can take a memory pool object in order to allocate
+from that memory pool specifically, however, if no memory pool object is
+provided there is a default memory pool which will be used.
+
+The memory pool introduced is a dynamic memory pool, as such no memory is
+immediately allocated on construction, instead memory is allocated by the OS
+for the pool when allocations are made and are freed back to the OS when the
+queue which enqueued the malloc and free commands is synchronized with.
```

Comment:
Any reason we have to be specific about the OS here that does the allocation, specifically when? or can we just say runtime? Ultimately yes the allocation will be done at the OS level, but the way it's phrased could prevent optimizations like caching on runtime level. This could just be implementation specific.

Reply:
Yeah that's a good point, I think we should rephrase this to say allocated by the SYCL RT.